tornado/iostream.py

```diff
@@ -415,6 +422,50 @@ def read_bytes(self, num_bytes, callback=None, streaming_callback=None,
             raise
         return future
 
+    def read_into(self, buf, callback=None, partial=False):
+        """Asynchronously read a number of bytes.
+
+        ``buf`` must be a writable buffer into which data will be read.
+        If a callback is given, it will be run with the number of read
+        bytes as an argument; if not, this method returns a `.Future`.
+
+        If ``partial`` is true, the callback is run as soon as any bytes
+        have been read.  Otherwise, it is run when the ``buf`` has been
+        entirely filled with read data.
+
+        .. versionadded:: 5.0
+        """
+        future = self._set_read_callback(callback)
+
+        # First copy data already in read buffer
+        available_bytes = self._read_buffer_size
+        n = len(buf)
+        if available_bytes >= n:
+            end = self._read_buffer_pos + n
+            buf[:] = memoryview(self._read_buffer)[self._read_buffer_pos:end]
+            del self._read_buffer[:end]
+            self._after_user_read_buffer = self._read_buffer
```

Comment:
The read is already satisfied here, wouldn't it be simpler to return instead of saving the current read buffer into `self._after_user_read_buffer`?

Reply:
I tried it both ways and this way seems cleaner to me. `_after_user_read_buffer` is a special case, but it lets us preserve the fact that `_run_read_callback` is the only place that cleans up after a read and sets results/runs callbacks. I think this one-to-one mapping between `read_*()` calls and `_run_read_callback` is helpful in reasoning about this class (and concretely, I'd rather not duplicate the logic around futures, callbacks, and the error listener)

Reply:
Then perhaps you can simply call `_run_read_callback` from here? AFAICT, this would make `_after_user_read_buffer` superfluous.

Reply:
The timing is tricky. `_run_read_callback` assigns `self._read_buffer = bytearray()`; we need to set it to something else. But by the time `_run_read_callback` returns, we've sent our results to the caller and they could have started another read, which would have incorrectly seen the empty read buffer. Or maybe this analysis is incorrect because callbacks are deferred to the next IOLoop iteration, but this is subtlety that I'd rather not rely on.

Reply:
Ok, fair enough.